File: app/src/main/java/com/mss/imagesearcher/presenter/detail/DetailPresenter.java
package com.mss.imagesearcher.presenter.detail;

import android.graphics.Bitmap;
import android.support.annotation.NonNull;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.mss.imagesearcher.domain.ImageListModel;
import com.mss.imagesearcher.view.detail.DetailView;

import java.io.FileNotFoundException;
import java.io.IOException;

import javax.inject.Inject;

import timber.log.Timber;

@InjectViewState
public class DetailPresenter extends MvpPresenter<DetailView> {

    private final ImageListModel model;

    @Inject
    public DetailPresenter(@NonNull final ImageListModel model) {
        this.model = model;
    }

    public void onCreate(int position) {
        Timber.d("onCreate %d", position);
        getViewState().showProgress(true);
        getViewState().showImage(model.getImages().get(position).getLargeImageURL());
    }

    public void onImageLoaded() {
        Timber.d("onImageLoaded");
        getViewState().showProgress(false);
    }

    public void onImageLoadFailed() {
        Timber.d("onImageLoadFailed");
        getViewState().showProgress(false);
    }

    public void onSaveClick(int position, @NonNull final Bitmap bitmap) {
        try {
            Timber.d("onSaveClick");
            model.saveBitmap(model.getImages().get(position), bitmap);
            getViewState().showSuccessSaveMessage();
        } catch (FileNotFoundException e) {
            Timber.e(e);
            getViewState().showFailedSaveMessage();
        } catch (IOException e) {
            Timber.e(e);
            getViewState().showFailedSaveMessage();
        }
    }

    public void onShareClick(int position, Bitmap bitmap) {
        try {
            Timber.d("onShareClick");
            String path = model.saveBitmap(model.getImages().get(position), bitmap);
            getViewState().shareImage(path);
        } catch (FileNotFoundException e) {
            Timber.e(e);
            getViewState().showFailedSaveMessage();
        } catch (IOException e) {
            Timber.e(e);
            getViewState().showFailedSaveMessage();
        }
    }
}

File: app/src/main/java/com/mss/imagesearcher/presenter/detail/DetailPresenter.kt
package com.mss.imagesearcher.presenter.detail

import android.graphics.Bitmap
import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.mss.imagesearcher.model.ImageListModel
import com.mss.imagesearcher.view.detail.DetailView
import timber.log.Timber
import java.io.FileNotFoundException
import java.io.IOException
import javax.inject.Inject

@InjectViewState
class DetailPresenter @Inject
constructor(val model: ImageListModel) : MvpPresenter<DetailView>() {

    init {
        model.currentImage.observeForever {
            viewState.apply {
                Timber.d("current image changed")
                showImage(false)
                model.currentImage.value?.let {
                    showProgress(true)
                    startLoadImage(it.largeImageURL)
                }
            }
        }
    }

    fun onImageLoaded() {
        Timber.d("onImageLoaded")
        viewState.showProgress(false)
        viewState.showImage(true)
    }

    fun onImageLoadFailed() {
        Timber.d("onImageLoadFailed")
        viewState.showProgress(false)
    }

    fun onSaveClick(bitmap: Bitmap) {
        try {
            Timber.d("onSaveClick")
            model.saveBitmap(model.images[0], bitmap)
            viewState.showSuccessSaveMessage()
        } catch (e: FileNotFoundException) {
            Timber.e(e)
            viewState.showFailedSaveMessage()
        } catch (e: IOException) {
            Timber.e(e)
            viewState.showFailedSaveMessage()
        }

    }

    fun onShareClick(bitmap: Bitmap) {
        try {
            Timber.d("onShareClick")
            val path = model.saveBitmap(model.images[0], bitmap)
            viewState.shareImage(path)
        } catch (e: FileNotFoundException) {
            Timber.e(e)
            viewState.showFailedSaveMessage()
        } catch (e: IOException) {
            Timber.e(e)
            viewState.showFailedSaveMessage()
        }
    }
}
